File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/CommissionsOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.utils.HypixelUtils;
import io.github.chindeaone.collectiontracker.utils.rendering.CommissionFormat;
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils;
import io.github.chindeaone.collectiontracker.utils.tab.CommissionsWidget;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class CommissionsOverlay implements AbstractOverlay{

    private final Position position = ConfigAccess.getCommissionsPosition();
    private final List<String> formattedCommissions = new ArrayList<>();
    private boolean renderingAllowed  = true;

    private List<String> getCommissionsLines() {
        List<String> raw = CommissionsWidget.INSTANCE.getRawCommissions();
        if (raw.isEmpty()) return Collections.emptyList();

        formattedCommissions.clear();
        CommissionFormat.Area detectedArea = null;

        for (String line : raw) {
            String formatted = line;
            String lowerLine = line.toLowerCase();
            for (CommissionFormat.CommissionType type : CommissionFormat.INSTANCE.getCOMMISSIONS()) {
                String typeNameLower = type.getName().toLowerCase();
                if (lowerLine.contains(typeNameLower)) {
                    formatted = type.getFormat().invoke(line);
                    if (detectedArea == null) detectedArea = type.getArea();
                    break;
                }
            }
            formattedCommissions.add(formatted);
        }

        if (detectedArea != null) {
            switch (detectedArea) {
                case DWARVEN_MINES -> formattedCommissions.addFirst("§2§l" + detectedArea.getDisplayName());
                case CRYSTAL_HOLLOWS -> formattedCommissions.addFirst("§5§l" + detectedArea.getDisplayName());
                case GLACITE_TUNNELS -> formattedCommissions.addFirst("§b§l" + detectedArea.getDisplayName());
            }
        }
        return formattedCommissions;
    }

    @Override
    public String overlayLabel() {
        return "Commissions Overlay";
    }

    @Override
    public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return ConfigAccess.isCommissionsEnabled() && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled()) return;
        List<String> lines = getCommissionsLines();

        if (lines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
            RenderUtils.renderStrings(context, lines)
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getCommissionsLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/CommissionsOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.ModLoader
import io.github.chindeaone.collectiontracker.config.commissionsPosition
import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.config.enableCommissionsOverlay
import io.github.chindeaone.collectiontracker.config.enableCommissionsTracking
import io.github.chindeaone.collectiontracker.tracker.commissions.CommissionsTracker.getCommissionsPerHour
import io.github.chindeaone.collectiontracker.tracker.commissions.CommissionsTracker.getCompletedCount
import io.github.chindeaone.collectiontracker.tracker.commissions.CommissionsTracker.getUptime
import io.github.chindeaone.collectiontracker.utils.parser.CommissionParser
import io.github.chindeaone.collectiontracker.utils.tab.CommissionWidget

class CommissionsOverlay : AbstractOverlay() {
    private var cachedLines: List<String> = emptyList()

    override val overlayLabel: String = "Commissions"

    override val position: Position get() = commissionsPosition

    override val isEnabled: Boolean get() = enableCommissionsOverlay

    override fun updateDimensions() {
        if (!isEnabled) return
        updateLinesIfNeeded()

        super.updateDimensions()
    }

    override val lines: List<String>
        get() {
            updateLinesIfNeeded()
            return cachedLines
        }

    private fun updateLinesIfNeeded() {
        if (!isEnabled) {
            cachedLines = emptyList()
            return
        }

        if (ModLoader.clientTicks % 5L != 0L) return

        val commissions = CommissionWidget.commissions
        if (commissions.isEmpty()) {
            cachedLines = emptyList()
            return
        }

        val isTracking = enableCommissionsTracking
        val completed = getCompletedCount()
        val perHour = getCommissionsPerHour()
        val currentUptime = getUptime()

        val newLines = mutableListOf<String>()
        var detectedArea: CommissionParser.Area? = null

        for (commission in commissions) {
            newLines.add(commission.formattedLine)
            if (detectedArea == null) {
                detectedArea = commission.type.area
            }
        }

        if (detectedArea != null) {
            when (detectedArea) {
                CommissionParser.Area.DWARVEN_MINES -> newLines.addFirst("§2§l" + detectedArea.displayName)
                CommissionParser.Area.CRYSTAL_HOLLOWS -> newLines.addFirst("§5§l" + detectedArea.displayName)
                CommissionParser.Area.GLACITE_TUNNELS -> newLines.addFirst("§b§l" + detectedArea.displayName)
            }
        }

        if (isTracking && completed > 0) {
            if (newLines.isNotEmpty()) {
                newLines.add("")
            }
            newLines.add("§6Commissions Completed: §e$completed")
            newLines.add("§6Commissions/h: §e" + String.format("%.2f", perHour))
            newLines.add("§6Uptime: §e$currentUptime")
        }

        cachedLines = newLines
    }
}
